ISO20022 PoC (#1940)
w3f/Grants-Program: applications/ISO20022.md

# ISO20022 PoC 

- **Team Name:** Open Smart Contract
- **Payment Address:** 3GTe5ArJLC8jesr6UFtB9kcoguQmvo8Cuo (BTC)

> - **[Level](https://github.com/w3f/Grants-Program/tree/master#level_slider-levels):** 3

Reviewer: ```suggestion
- **[Level](https://github.com/w3f/Grants-Program/tree/master#level_slider-levels):** 2
```

## Project Overview 

Response to [RFP 20022](https://github.com/w3f/Grants-Program/blob/master/docs/RFPs/Open/ISO_20022.md)

> Project name: ISO20022 PoC
> 

Reviewer: ```suggestion
```

I think you can skip this here since it's already in the title.

> Cross border payment example as proof of concepts,leveraging the unique [Off-Chain Features of Substrate](https://docs.substrate.io/v3/concepts/off-chain-features/) that show the advantages of using [ISO20022](https://www.iso20022.org/) together with Substrate. 

Reviewer: ```suggestion
Cross-border payment example as proof of concepts, leveraging the unique [Off-Chain Features of Substrate](https://docs.substrate.io/v3/concepts/off-chain-features/) that show the advantages of using [ISO20022](https://www.iso20022.org/) together with Substrate. 
```

	- ISO20022 messaging and payment events and stats 
	- ISO20022 messaging and payment queries 
- PoC/MVP or other relevant prior work or research on the topic
	- Our research shows that multiple gaps exist to support ISO20022 compliant cross border payment use case  
		- There's big ata type compatibility gaps between Polkadot and IS20022, such as date and time, currency, float, array, etc. 
		- There's no out of box support of SO20022 XML processing and mapping such as XSD transformations 
		- There's no built-in data validation, message routing, payment securties, etc.  
		- There's no native foreign exchanage / currency conversion capabilities
		- There's no payment states reflecting current cross border payment flow states 
		
- What your project is _not_ or will _not_ provide or implement
  - This is not a complete cross border payment system given the scope of proof of concepts 
  - Likely this is the first ISO20022 project on Dotsama, there could be unexpected gaps discovered
  - This project is not a full blown IS20022 implementation on Dotsama / Substrate but explore to expose gaps for future endeavors    

>   - This project is not about payment identity / KYC / AML solutions 

Reviewer: To increase the readability of this section, I think it would be nice if you could write a text on what you're planning to implement and why, rather than just having these bullet points.

Author: We are following the application template here. Happy to write a paragraph to summarize. Again this milestone is not about implementation but specifying what to implement.

- What need(s) does your project meet?
	- Explore possibilities and gaps of Polkadot / Substrate support of ISO20022 standard for future improvements 
- Are there any other projects similar to yours in the Substrate / Polkadot / Kusama ecosystem?
  - If so, how is your project different?
  - If not, are there similar projects in related ecosystems? There are known blockchains already claim to support ISO20020 such as Ripple XRP, XLM, Algorand, etc. 

## Team 

### Team members

- Name of team leader: Daniel Deng

### Contact

- **Contact Name:** Daniel Deng
- **Contact Email:** OpenSmartContract@gmail.com

### Legal Structure

- **Registered Address:** No registred entity
- **Registered Legal Entity:** No registred entity

### Team's experience

Daniel is the Chief Product Officer at HeroDAO where he has led blockchain based smart contract products and development on Polkodot / Kusama blockchains. Previous at Meta he managed crypto (Libra / Diem) and fiat payments and web3 initiatives. Prior to that, he was Senior Director of Product Management at Visa managing Visa's global payment product portfolios and platforms.

> Here's a recent[ Medium article](https://medium.com/@opensmartcontract/learn-ink-by-example-order-food-on-blockchain-a4024b2dee4a) on ink! smart contract with Rust macros implementing dApp on Dotsama chain, with [Smart Contract Deployment and Interaction Guide using Contracts-UI and Polkadot.js](https://github.com/InkSmartContract/BlockchainFoodOrder/blob/main/BlockchainFoodOrder%20Smart%20Contract%20Deployment%20and%20Interactions%20Guide.pdf).        

Reviewer: ```suggestion
Here's a recent [Medium article](https://medium.com/@opensmartcontract/learn-ink-by-example-order-food-on-blockchain-a4024b2dee4a) on ink! smart contract with Rust macros implementing dApp on Dotsama chain, with [Smart Contract Deployment and Interaction Guide using Contracts-UI and Polkadot.js](https://github.com/InkSmartContract/BlockchainFoodOrder/blob/main/BlockchainFoodOrder%20Smart%20Contract%20Deployment%20and%20Interactions%20Guide.pdf).        
```

### Team Code Repos
- [Order Food on the Blockchain](https://github.com/InkSmartContract/BlockchainFoodOrder)

Please also provide the GitHub accounts of all team members. 
- [GitHub1](https://github.com/OpenSmartContract) 
- [Github2](https://github.com/FPLeader) 

### Team LinkedIn Profiles (if available)
- [Profile](https://www.linkedin.com/in/dengdaniel)

## Development Status 

- [Blockchain Food Order dApp](https://github.com/InkSmartContract/BlockchainFoodOrder/blob/main/Learn%20ink!%20by%20Example%20%E2%80%94%20Order%20Food%20on%20the%20Blockchain.pdf)
- [Smart Contract Deployment and Interaction Guide using Contracts-UI and Polkadot.js](https://github.com/InkSmartContract/BlockchainFoodOrder/blob/main/BlockchainFoodOrder%20Smart%20Contract%20Deployment%20and%20Interactions%20Guide.pdf)

## Development Roadmap 

### Overview

- **Estimated duration:** ~1-2 months
- **FTE:**  1,2
- **Costs:** 25K USDd

### Milestone 1

- **Estimated duration:** ~1-2 months
- **FTE:**  1,2
- **Costs:** 25K USD

| Number | Deliverable | Specification |
| -----: | ----------- | ------------- |
| **0a.** | License | Apache 2.0 |

> | **0b.** | Documentation | Design documents with details on what it takes for Polkadot/Substrate to support ISO20022 messages, gaps and solution recommendations |

Reviewer: Could you add the [mandatory deliverables](https://github.com/w3f/Grants-Program/blob/19d2f54c0cd323d55518910aabfb88a9f407f677/applications/application-template.md#milestone-1-example--basic-functionality)?

Author: The deliverable for this milestone is the use case definition, gap analysis, implementation specification, etc. as listed. Implementation in potential future milestones.

> | 1. | Design the use case | Degine an ISO20022 cross border payment use case with persona, end-to-end flow | 

Reviewer: This should be part of the proposal. We typically only cover the implementation itself and would expect grantees to show us sufficient details of what they want to implement. Using an "e2e flow" (e.g. sequence diagram) certainly would be useful for that.

Author: This milestone is all about design / specification, not implementations. Noted that e2e flow would be part of the deliverables of this milestone.

> | 2. | Specify ISO20022 messages | Specify ISO20022 messages for each and every step of the cross border payment flow, including reference data fields besides core XML data | 

Reviewer: This should be part of the proposal. We typically only cover the implementation itself and would expect grantees to show us sufficient details of what they want to implement. Using an "e2e flow" (e.g. sequence diagram) certainly would be useful for that.

Author: same as above

> | 3. | ISO20022 XML <> Polkadot XCM data mapping | Identify data types gaps, XML<>XCM trnasformation and validations for the use case| 

Reviewer: ```suggestion
| 3. | ISO20022 XML <> Polkadot XCM data mapping | Using data type gaps as an input, we will implement XML<>XCM transformations and validations | 
```

> for the use case

Do you mean some specific use case related to ISO20022 here? Otherwise I think you can just skip this part.

Author: Yes for the specific cross-border payment use case in this example related to ISO20022.

> | 4. | ISO20022 <> Polkadot messaging | Identify XCMP gaps in messaging, relaying, routing feasibilities for the use case | 

Reviewer: What's the difference to the previous deliverable here?

Author: Are you referring to #4 here vs #3 here?

Reviewer: Yes, exactly. As far as I understand ISO20022 and XCM are message formats. However, XCMP is the transport mechanism. I believe ISO20022 is agnostic to the underlying transport mechanism, right?

> | 5. | Define offchain data storage | Define offchain data structure, storage and indexes for ISO20022 messages and payments in this use case |

Reviewer: ```suggestion
| 5. | Define offchain data storage | Implement offchain data structure, storage and indexes for ISO20022 messages and payments in this use case |
```

Author: Same clarification. This milestone is all about design and specifications, not implementations.

> | 6. | Define offchian worker module | Define OCW logic to handle ISO20022 messages mapping, processing, messaging in this use case |

Reviewer: ```suggestion
| 6. | Define offchian worker module | Implement OCW logic to handle ISO20022 messages mapping, processing, messaging in this use case |
```

> | 7. | Define Smart contracts | Define ink! smart contracts to implement X-border payments and offchian integrations in this use case |

Reviewer: ```suggestion
| 7. | Define Smart contracts | Implement ink! smart contracts to support X-border payments and off chain integrations in this use case |
```

Author: Same clarification. This milestone is all about design and specifications, not implementations.

> | 8. | Define ISO20022 messaging and payment events and stats | Define ISO20020 messaging and payment flow events and states |

Reviewer: ```suggestion
| 8. | Define ISO20022 messaging and payment events and stats | Implement ISO20020 messaging and payment flow events and states |
```

Author: Same clarification. This milestone is all about design and specifications, not implementations.

> | 9. | Define ISO20022 messaging and payment queries | Define how to query ISO20022 messaging and payments, based off events, stats, transactions and indexes stored offchain |

Reviewer: This deliverable would be related to querying and watching ISO20022 messages, is that correct?
Please also make it clear here that you're implementing something.

Author: Yes define how ISO20022 messages and status can be queried, not implementation in this milestone.

## Future Plans

- Publish learning and share with the community, similar to this [ Medium article](https://medium.com/@opensmartcontract/learn-ink-by-example-order-food-on-blockchain-a4024b2dee4a) and/or [Contracts-UI and Polkadot.js Smart Contract Deployment and Interaction Guide](https://github.com/InkSmartContract/BlockchainFoodOrder/blob/main/BlockchainFoodOrder%20Smart%20Contract%20Deployment%20and%20Interactions%20Guide.pdf)

> - Long term we intend to expand this ISO20022 PoC to:

Reviewer: How many people are going to work on this project?

Reviewer: That's fine. However, those consultants should be listed in the team section as well. If there changes down the road, you'll be able to amend it later, even after the proposal would be approved.

Author: Noted in [revised proposal ](https://github.com/w3f/Grants-Program/pull/1940/commits/a17d45e75dda9d6c465488ce454fdb8e0446fc16)

# ISO20022 PoC 

- **Team Name:** Open Smart Contract

> - **Payment Address:** 3GTe5ArJLC8jesr6UFtB9kcoguQmvo8Cuo

Reviewer 2: ```suggestion
- **Payment Address:** 3GTe5ArJLC8jesr6UFtB9kcoguQmvo8Cuo (BTC)
```

Author: Updated in this [commit](https://github.com/OpenSmartContract/W3F-Program/commit/16385bba4c45b5ba0dcd246542d4ba586e5e93c6)

> - **[Level](https://github.com/w3f/Grants-Program/tree/master#level_slider-levels):** 2

Reviewer 2: ```suggestion
- **[Level](https://github.com/w3f/Grants-Program/tree/master#level_slider-levels):** 3
```

- Who is your target audience (parachain/dapp/wallet/UI developers, designers, your own user base, some dapp's userbase, yourself)?
	- Targeting ecosystem develpopers to levearage on the learning and tools to further develop ISO20022 compitable dApps, to make it possible for transitional financial institutions to interoprate with Polkadot / Kusama / Substrate blockchains based payment apps. 
- What need(s) does your project meet?
	- Explore possibilities and gaps of Polkadot / Substrate support of ISO20022 standard for future improvements 
- Are there any other projects similar to yours in the Substrate / Polkadot / Kusama ecosystem?
  - If so, how is your project different?
  - If not, are there similar projects in related ecosystems? There are known blockchains already claim to support ISO20020 such as Ripple XRP, XLM, Algorand, etc. 

## Team 

### Team members

- Name of team leader: Daniel Deng

### Contact

- **Contact Name:** Daniel Deng
- **Contact Email:** OpenSmartContract@gmail.com

### Legal Structure

- **Registered Address:** No registred entity

> - **Registered Legal Entity:** No registred entity

Reviewer 2: Please provide at least an address. If you prefer not to do so publicly, feel free to [email](mailto:grants@web3.foundation) it with reference to this pull request.

Author: Will email with reference to this PR

Here's a recent[ Medium article](https://medium.com/@opensmartcontract/learn-ink-by-example-order-food-on-blockchain-a4024b2dee4a) on ink! smart contract with Rust macros implementing dApp on Dotsama chain, with [Smart Contract Deployment and Interaction Guide using Contracts-UI and Polkadot.js](https://github.com/InkSmartContract/BlockchainFoodOrder/blob/main/BlockchainFoodOrder%20Smart%20Contract%20Deployment%20and%20Interactions%20Guide.pdf).        

### Team Code Repos
- [Order Food on the Blockchain](https://github.com/InkSmartContract/BlockchainFoodOrder)

Please also provide the GitHub accounts of all team members. 
- [GitHub1](https://github.com/OpenSmartContract) 
- [Github2](https://github.com/FPLeader) 

### Team LinkedIn Profiles (if available)
- [Profile](https://www.linkedin.com/in/dengdaniel)

## Development Status 

> - [RFPs](https://github.com/w3f/Grants-Program/tree/master/docs/RFPs) (requests for proposal)

Reviewer 2: ```suggestion
```

- [Blockchain Food Order dApp](https://github.com/InkSmartContract/BlockchainFoodOrder/blob/main/Learn%20ink!%20by%20Example%20%E2%80%94%20Order%20Food%20on%20the%20Blockchain.pdf)
- [Smart Contract Deployment and Interaction Guide using Contracts-UI and Polkadot.js](https://github.com/InkSmartContract/BlockchainFoodOrder/blob/main/BlockchainFoodOrder%20Smart%20Contract%20Deployment%20and%20Interactions%20Guide.pdf)

## Development Roadmap 

### Overview

- **Total Estimated Duration:** ~3-6 months
- **Full-Time Equivalent (FTE):**  1,4 FTE

> - **Total Costs:** 3 BTC 

Reviewer 2: The milestone and total costs should be denominated in USD. We will do the conversion to BTC at time of payment.

Author: Updated in this [commit](https://github.com/OpenSmartContract/W3F-Program/commit/16385bba4c45b5ba0dcd246542d4ba586e5e93c6)

> ### Milestone 1 Example — Basic functionality

Reviewer 2: Please replace template text

Author: Replaced template text in this [commit](https://github.com/OpenSmartContract/W3F-Program/commit/16385bba4c45b5ba0dcd246542d4ba586e5e93c6)

| 2. | Specify ISO20022 messages | Specify ISO20022 messages for each and every step of the cross border payment flow, including reference data fields besides core XML data | 
| 3. | ISO20022 XML <> Polkadot XCM data mapping | Identify data types gaps, XML<>XCM trnasformation and validations for the use case| 
| 4. | ISO20022 <> Polkadot messaging | Identify XCMP gaps in messaging, relaying, routing feasibilities for the use case | 
| 5. | Define offchain data storage | Define offchain data structure, storage and indexes for ISO20022 messages and payments in this use case |
| 6. | Define offchian worker module | Define OCW logic to handle ISO20022 messages mapping, processing, messaging in this use case |
| 7. | Define Smart contracts | Define ink! smart contracts to implement X-border payments and offchian integrations in this use case |
| 8. | Define ISO20022 messaging and payment events and stats | Define ISO20020 messaging and payment flow events and states |
| 9. | Define ISO20022 messaging and payment queries | Define how to query ISO20022 messaging and payments, based off events, stats, transactions and indexes stored offchain |

> ### Milestone 2 Example — Additional features

Reviewer 2: Please replace template text

Author: Replaced template text in this [commit](https://github.com/OpenSmartContract/W3F-Program/commit/16385bba4c45b5ba0dcd246542d4ba586e5e93c6)
